source/common/common/utility.cc

```diff
@@ -122,6 +123,15 @@ std::vector<std::string> StringUtil::split(const std::string& source, const std:
   return ret;
 }
 
+std::string StringUtil::join(const std::vector<std::string>& source, const std::string& delimiter) {
+  std::ostringstream buf;
+  std::copy(source.begin(), source.end(),
+            std::ostream_iterator<std::string>(buf, delimiter.c_str()));
+  std::string ret = buf.str();
+  // copy will always end with an extra delimiter, we remove it here.
+  return ret.substr(0, ret.length() - delimiter.length());
```

Comment:
What happens if the list of elements to join is empty, the count is then negative, is `substr` guaranteed to behave as if it is zero here?